distributed/worker.py

```diff
@@ -3089,7 +3093,8 @@ async def gather_dep(
                         self.batched_stream.send(
                             {"op": "missing-data", "errant_worker": worker, "key": d}
                         )
-                        recommendations[ts] = "fetch" if ts.who_has else "missing"
+                        if ts.state != "released":
+                            recommendations[ts] = "fetch" if ts.who_has else "missing"
```

Comment:
We should not execute this entire clause now that we're dealing with the OSError above separately. Should also be something like

`elif ts not in recommendations and ts.state in {"fetch", "flight"}:`

Reply:
This is very brittle. I hate it. We only want to execute parts of the finally clause...

Reply:
another "brittle an very ugly" fix might be to create a set in the os error handler to remember the tasks we don't want to be handled here.

Reply:
`ts not in recommendations` is already basically handled a few lines above: https://github.com/dask/distributed/pull/6123/files#diff-0c7375c95aaa92986f8ed5460bae4e98720a14fca13556e01fd44ab6b927c24dR3089

Doesn't this clause only come into play if the `get-data` request succeeds, but the worker just doesn't have the key?

I don't understand how the task could be released here. I'm curious if just https://github.com/dask/distributed/pull/6123/files#r850533377 would fix things without your change here.

Reply:
```python
elif ts not in recommendations and ts.state in {"fetch", "flight"}:
```

This change would cause a few tests to fail, one among them is `distributed/tests/test_worker.py::test_task_flight_compute_oserror`

Reply:
> I don't understand how the task could be released here. I'm curious if just https://github.com/dask/distributed/pull/6123/files#r850533377 would fix things without your change here.

This doesn't seem to point to anything in particular.  Maybe the link was broken by a push?